File: pset1/credit/credit.c
#include <stdio.h>
#include <cs50.h>

int main(void)
{
    long number;
    do
    {
        number = get_long("Number: ");
    }
    while (number <= 0);
    long number_sum1 = number;
    int sum_1 = 0;
    int sum_2 = 0;
    int rest_1 = 0;
    int rest_2 = 0;
    for (long i = 100; i < 10 * number_sum1; i = i * 100)
    {
        rest_1 = (number_sum1 % i - number_sum1 % (i / 10)) / (i / 10);
        if (2 * rest_1 > 9)
        {
            sum_1 = sum_1 + 2 * rest_1 - 9;
        }
        else
        {
            sum_1 = sum_1 + 2 * rest_1;  
        }
    }
    for (long j = 10; j < 10 * number_sum1; j = j * 100)
    {
        rest_2 = (number_sum1 % j - number_sum1 % (j / 10)) / (j / 10);
        sum_2 = sum_2 + rest_2;
    }
    int sum = sum_1 + sum_2;
    if (sum % 10 == 0)
    {
        if (number >= 1000000000000000)
        {
            if (number < 10000000000000000)
            {
                if (number >= 5100000000000000 && number < 5600000000000000)
                {
                    printf("MASTERCARD\n"); 
                }
                else if (number >= 4000000000000000 && number < 5000000000000000)
                {
                    printf("VISA\n");
                }
                else
                {
                    printf("INVALID\n");
                }
            }
            else
            {
                printf("INVALID\n"); 
            }
        }
        else if ((number >= 340000000000000 && number < 350000000000000) || (number >= 370000000000000 && number < 380000000000000))
        {
            printf("AMEX\n");
        }
        else if (number >= 4000000000000)
        {
            if (number < 5000000000000)
            {
                printf("VISA\n");
            }
            else
            {
                printf("INVALID\n");
            }
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
}

File: pset6/credit/credit.py
from cs50 import get_int


def luhns_algorithm(number):
    number_list = list(map(int, str(number)))
    sum_luhn = 0
    for i in range(1, len(number_list) + 1):
        if i % 2 != 0:
            sum_luhn += number_list[-i]
        else:
            if number_list[-i] > 4:
                sum_luhn += 2 * number_list[-i] - 9
            else:
                sum_luhn += 2 * number_list[-i]
        i += 1
    if sum_luhn % 10 == 0:
        return True
    else:
        return False


def issuer(number):
    if number in range(4 * 10 ** 12, 5 * 10 ** 12) or number in range(4 * 10 ** 15, 5 * 10 ** 15):  # 13 or 16 digit (visa)
        return "VISA"
    elif number in range(34 * 10 ** 13, 35 * 10 ** 13) or number in range(37 * 10 ** 13, 38 * 10 ** 13):  # 15 digit (amex)
        return "AMEX"
    elif number in range(51 * 10 ** 14, 56 * 10 ** 14):  # 16 digit (master)
        return "MASTERCARD"
    else:
        return "INVALID"


def main():
    while True:
        number = get_int("Number: ")
        if number > 0:
            break
    if luhns_algorithm(number) == True:
        print(issuer(number))
    else:
        print("INVALID")


main()
